[New] Apply RGB renderer

yo1995 opened the pull request:
## Description

This PR implements `Apply RGB renderer` in `Visualization` category.
[URL to README](https://github.com/Esri/arcgis-maps-sdk-swift-samples/blob/a159ac2c256c47984d552b719f672f39050cfc5c/Shared/Samples/Apply%20RGB%20renderer/README.md)

## Linked Issue(s)

- `swift/issues/6850`
- `native-apis-doc/issues/7768`

https://desktop.arcgis.com/en/arcmap/latest/manage-data/raster-and-images/stretch-function.htm

## How To Test

Run the sample, choose different stretch rendering options.

## Screenshots

![image](https://github.com/user-attachments/assets/4176e90a-ebcc-447c-9b8d-8d510d48f052)

## To Discuss

- What do you think about the range slider? A different UI can be used for range selection. The cocoa implementation is [here](https://github.com/Esri/arcgis-runtime-samples-ios/tree/main/arcgis-ios-sdk-samples/Layers/RGB%20renderer) with 3 textfields.

yo1995 commented on Shared/Samples/Apply RGB renderer/README.md:
```diff
@@ -0,0 +1,37 @@
+2. Create a `RasterLayer` from the raster.
+3. Create a `Basemap` from the raster layer and set it to the map.
+4. Create an `RGBRenderer`, specifying the `StretchParameters` and other properties.
+5. Set the `Renderer` on the raster layer with `rasterLayer.setRenderer(renderer)`.
```
```suggestion
5. Set the `Renderer` on the raster layer.
```

dfeinzimer commented:
> What do you think about the range slider? A different UI can be used for range selection. The cocoa implementation is [here](https://github.com/Esri/arcgis-runtime-samples-ios/tree/main/arcgis-ios-sdk-samples/Layers/RGB%20renderer) with 3 textfields.

Did you consider using a slider for Standard Deviation mode?

dfeinzimer commented:
Looks like there's a current merge conflict.

des12437 commented on Shared/Samples/Apply RGB renderer/ApplyRGBRendererView.swift:
```diff
@@ -0,0 +1,70 @@
+import ArcGIS
+import SwiftUI
+
+struct ApplyRGBRendererView: View {
```
A few comments:
- The map can be declared rather than set in a task.
- The raster layer does not need to be loaded, it can be loaded implicitly when the map loads.
- The raster layer can be a computed property.
- Since the layer is not loaded, the error logic can be removed.
- Why is the viewpoint set? The map extent looks identical when the viewpoint is not set.

des12437 commented on Shared/Samples/Apply RGB renderer/ApplyRGBRendererView.swift:
```diff
@@ -0,0 +1,70 @@
+import SwiftUI
+
+struct ApplyRGBRendererView: View {
+    /// An empty map.
+    @State private var map = Map()
```
```suggestion
    /// A map with a raster layer.
    @State private var map: Map = {
        let raster = Raster(
            fileURL: Bundle.main.url(
                forResource: "Shasta",
                withExtension: "tif",
                subdirectory: "raster-file/raster-file"
            )!
        )
        let rasterLayer = RasterLayer(raster: raster)
        return Map(basemap: Basemap(baseLayer: rasterLayer))
    }()
```

des12437 commented on Shared/Samples/Apply RGB renderer/ApplyRGBRendererView.swift:
```diff
@@ -0,0 +1,70 @@
+    /// The error shown in the error alert.
+    @State private var error: Error?
+    
+    /// The raster layer to apply RGB renderer.
+    @State private var rasterLayer: RasterLayer?
```
```suggestion
    /// The raster layer to apply RGB renderer.
    private var rasterLayer: RasterLayer {
        map.basemap?.baseLayers[0] as! RasterLayer
    }
```

This can be a computed property.

des12437 commented on Shared/Samples/Apply RGB renderer/ApplyRGBRendererView.swift:
```diff
@@ -0,0 +1,70 @@
+                    .disabled(rasterLayer == nil)
+                }
+            }
+            .task {
+                do {
+                    let raster = Raster(
+                        fileURL: Bundle.main.url(
+                            forResource: "Shasta",
+                            withExtension: "tif",
+                            subdirectory: "raster-file/raster-file"
+                        )!
+                    )
+                    let rasterLayer = RasterLayer(raster: raster)
+                    self.rasterLayer = rasterLayer
+                    try await rasterLayer.load()
+                    map = Map(basemap: Basemap(baseLayer: rasterLayer))
+                    if let extent = rasterLayer.fullExtent {
+                        map.initialViewpoint = Viewpoint(boundingGeometry: extent)
+                    }
+                } catch {
+                    self.error = error
+                }
+            }
```
```suggestion
```

des12437 commented on Shared/Samples/Apply RGB renderer/ApplyRGBRendererView.swift:
```diff
@@ -0,0 +1,70 @@
+                        .presentationDetents([.fraction(0.5)])
+                        .frame(idealWidth: 320, idealHeight: 360)
+                    }
+                    .disabled(rasterLayer == nil)
```
```suggestion
```
If the raster layer is a computed property this check can be removed.

des12437 reviewed (commented):
The renderer settings stay applied to the map when the settings sheet is closed, but when the sheet is opened again the settings applied to the map are cleared.  Is this expected? The 100.x sample does not have this behavior.

yo1995 commented:
> Did you consider using a slider for Standard Deviation mode?

@dfeinzimer Thanks for the suggestion - I changed the textfield to a slider.

> The renderer settings stay applied to the map when the settings sheet is closed, but when the sheet is opened again the settings applied to the map are cleared. Is this expected?

@des12437 I've moved the state property `stretchType` to the parent view so it persists while the sheet is dismissed.

des12437 commented on Shared/Samples/Apply RGB renderer/ApplyRGBRendererView.SettingsView.swift:
```diff
@@ -0,0 +1,149 @@
+        /// The raster layer to apply the RGB renderer to.
+        let rasterLayer: RasterLayer
+        
+        /// The stretch type to apply to the raster layer.
+        @Binding var stretchType: StretchType
+        /// The color to use for the minimum values in the min-max stretch.
+        @State private var minColor: Color = .black
+        /// The color to use for the maximum values in the min-max stretch.
+        @State private var maxColor: Color = .green
+        /// The minimum value for the percent clip stretch.
+        @State private var minValue: Double = 20
+        /// The maximum value for the percent clip stretch.
+        @State private var maxValue: Double = 80
+        /// The standard deviation factor for the standard deviation stretch.
+        @State private var standardDeviation = 0.5
```
```suggestion
        /// A binding to the renderer settings.
        @Binding var settings: ApplyRGBRendererView.RendererSettings
```

des12437 commented on Shared/Samples/Apply RGB renderer/ApplyRGBRendererView.SettingsView.swift:
```diff
@@ -0,0 +1,149 @@
+                        }
+                    }
+                }
+            }
```
```suggestion
            }
            .onChange(of: settings, initial: true) {
                switch settings.stretchType {
                case .histogramEqualization:
                    // Nothing to configure for histogram equalization.
                    // It is useful when there are a lot of pixel values
                    // that are closely grouped together
                    setStretchParameters(HistogramEqualizationStretchParameters())
                case .minMax:
                    // The pixel values which serve as endpoints for
                    // the histogram used for the stretch.
                    setStretchParameters(MinMaxStretchParameters(minValues: settings.minColor.rgb, maxValues: settings.maxColor.rgb))
                case .percentClip:
                    // The percentile cutoff above which pixel values
                    // in the raster dataset are to be clipped.
                    setStretchParameters(PercentClipStretchParameters(min: settings.minValue, max: settings.maxValue))
                case .standardDeviation:
                    // It applies a linear stretch between the
                    // values defined by the standard deviation.
                    setStretchParameters(StandardDeviationStretchParameters(factor: settings.standardDeviation))
                }
            }
```
The onChange view modifiers can be consolidated into one modifier.

des12437 commented on Shared/Samples/Apply RGB renderer/ApplyRGBRendererView.swift:
```diff
@@ -0,0 +1,59 @@
+        return Map(basemap: Basemap(baseLayer: RasterLayer(raster: raster)))
+    }()
+    
+    /// The stretch type to apply to the raster layer.
+    @State private var stretchType: StretchType = .histogramEqualization
```
```suggestion
    /// The settings for a RGB renderer.
    struct RendererSettings: Equatable {
        /// The stretch type to apply to the raster layer.
        var stretchType: StretchType = .histogramEqualization
        /// The color to use for the minimum values in the min-max stretch.
        var minColor: Color = .black
        /// The color to use for the maximum values in the min-max stretch.
        var maxColor: Color = .green
        /// The minimum value for the percent clip stretch.
        var minValue: Double = 20
        /// The maximum value for the percent clip stretch.
        var maxValue: Double = 80
        /// The standard deviation factor for the standard deviation stretch.
        var standardDeviation = 0.5
    }
    
    /// The RGB renderer settings.
    @State private var rendererSettings = RendererSettings()
```

yo1995 replied:
Thanks, great suggestion. I've refactored a bit.

des12437 reviewed (commented):
>The renderer settings stay applied to the map when the settings sheet is closed, but when the sheet is opened again the settings applied to the map are cleared. Is this expected?

>>@des12437 I've moved the state property stretchType to the parent view so it persists while the sheet is dismissed.

The stretch type persists but not the other settings, such as the min/max color. Another option is to add a binding for a renderer settings property so all settings persist.

des12437 reviewed (approved):
Looks good, thanks for making those changes.

dfeinzimer commented on Shared/Samples/Apply RGB renderer/ApplyRGBRendererView.SettingsView.swift:
```diff
@@ -0,0 +1,70 @@
+                        RangeSlider(lowerValue: $settings.minValue, upperValue: $settings.maxValue, range: 0...100, step: 1)
+                    case .standardDeviation:
+                        LabeledContent("Factor", value: settings.standardDeviation, format: .number.precision(.fractionLength(2)))
+                        Slider(value: $settings.standardDeviation, in: 0.0...16.0, step: 0.01) {
```
This value doesn't seem to make any noticeable difference when it's > ~6.0? Just wondering why allow all the way up to 16?

yo1995 replied:
I had a bit struggle when picked the upper bound.
For a 0 - 255 pixel band value, the biggest possible standard deviation is sqrt(255) = 16. But for most images, since the values will be closer to the center of the histogram, the std will be much smaller (for example, std = 10 can cover a span of 32 - 128). Still, for the most extreme case (all bands are concentrated at either end of 0 or 255), an upper bound 16 will be sufficient.

yo1995 commented:
Thanks for the feedback!
